diff --git a/src/ConnectionXTRX/CMakeLists.txt b/src/ConnectionXTRX/CMakeLists.txt
index 689b2f45..b5c5ce0e 100644
--- a/src/ConnectionXTRX/CMakeLists.txt
+++ b/src/ConnectionXTRX/CMakeLists.txt
@@ -43,11 +43,11 @@ find_library(LITEPCIE_LIBRARY litepcie
 ########################################################################
 ## CUDA
 ########################################################################
-find_package(CUDAToolkit)
+#find_package(CUDAToolkit)
 
 ########################################################################
 ## Add to library
 ########################################################################
 target_include_directories(LimeSuite PUBLIC ${LITEPCIE_INCLUDE_DIR})
 target_sources(LimeSuite PRIVATE ${CONNECTION_XTRX_SOURCES})
-target_link_libraries(LimeSuite ${LITEPCIE_LIBRARY} crypt CUDA::cuda_driver)
+target_link_libraries(LimeSuite ${LITEPCIE_LIBRARY} crypt)
